Interoperable DNS Server Cookies support

wtoorop opened the pull request.

wtoorop committed:
DNS Coookies support

wtoorop committed:
Rename do_answer_cookie in answer_cookie for config

wtoorop committed:
SipHash2.4 takes 128 bit keys!

wtoorop committed:
Merge branch 'master' into features/cookies

wtoorop committed:
Server cookies up-to-date with draft-sury-toorop-dnsop-server-cookies-01

wtoorop committed:
Siphash is only alg which has 128 bit secret

wtoorop committed:
Don't overwrite what we have to compare

wtoorop committed:
Merge branch 'master' into features/cookies

wtoorop committed:
Small fixes

wtoorop committed:
Merge branch 'master' into features/cookies

wtoorop committed:
Adapt cutest for query_create() and query_add_optional()

wtoorop committed:
Fix checkconf test for "answer-cookie: yes" default config

Commit:
{set,del}_cookie stubs

Commit:
dns-cookies: support for kind of cookie secret history

Commit:
dns-cookies: fix cookie secret size

Commit:
dns-cookies: provide push_cookie, drop_cookie, show_cookies commands

Commit:
dns-cookies: add newline to error message

Commit:
dns-cookies: fix check against secret

Commit:
dns-cookies: add cookie specific config options; extend lexer and grammer accordingly

Commit:
dns-cookies: add tasks for push and drop

Commit:
dns-cookies: add dump cookie secret function

Commit:
dns-cookies: remote: call corresponding tasks

Commit:
dns-cookies: print_cookie_secrets cleanup

Commit:
dns-cookies: fix nsd-control argument parsing

Commit:
dns-cookies: reload cookie store file on startup

Commit:
dns-cookie: minor comment on zero termination

Commit:
dns-cookies: remove timestamps from secrets

Commit:
dns-cookies: actually dispatch cookie secret tasks

Commit:
dns-cookies: c89 compat

wtoorop committed:
Merge pull request #135 from magenbluten/feature-cookies

Feature cookies

wtoorop committed:
Merge branch 'master' into features/cookies

wtoorop commented:
Copy TODO from @magenbluten's branch:

- [x] unit tests

wtoorop committed:
Remove leftover conflict marker

wtoorop committed:
Get rid of C90 warnings

wtoorop committed:
Reuse cookies that are valid within 30 minutes

Commit:
first cookies test tpkg

Commit:
change test to listen to all loopback interface addresses

Commit:
rewrite flow to incorporate shell opening

Commit:
add A2 test

Commit:
add A3 test

Commit:
fix typo

Commit:
correct a3 test and add teardown

Commit:
add a4

Commit:
execute a4

Commit:
fix a4 dig

Commit:
fix a4 and output text in a*

wtoorop committed:
Fix Cookie Reuse

wtoorop committed:
Fix, simplify and speedup dns-cookies test

wtoorop committed:
Rename cookiesecretfile to cookie-secret-file

wtoorop committed:
Allow getting and printing of cookie options

wtoorop committed:
Changes to allow safe cookie secret roll-overs

Cookie secrets now have a status.
  - Active secrets will be used for creating new DNS Cookies
  - Staging secrets can only be used for verifying DNS Cookies

Change the name from nsd-control command push_cookie_secret to add_cookie_secret
  - If there are no cookie secrets, add_cookie_secret will add an active cookie secret
  - Otherwise it will add (or replace) the staging cookie secret.

Add nsd-control command activate_cookie_secret.
  - activate_cookie_secret will activate the staging cookie secret and make the active one staging

wtoorop committed:
Update documentation on cookie secret manipulation

wtoorop committed:
Compare DNS Cookie hashes with CRYPTO_memcmp

wtoorop committed:
Merge branch 'master' into features/cookies

wtoorop committed:
cookie-secret-file option may be unset

wtoorop committed:
Merge branch 'master' into features/cookies

wtoorop committed:
Fix checkconf test with cookie-secret-file options

wcawijngaards commented on server.c:
```diff
@@ -3580,6 +3583,7 @@ handle_tcp_reading(int fd, short event, void* arg)
 	ssize_t received;
 	struct event_base* ev_base;
 	struct timeval timeout;
+	uint32_t now;
```
```suggestion
	uint32_t now = 0;
```
Perhaps this is useful

wcawijngaards commented on util.c:
```diff
@@ -1234,3 +1235,58 @@ int set_cpu_affinity(cpuset_t *set)
+	if(nsd->cookie_count == 0) {
+		memcpy( nsd->cookie_secrets->cookie_secret
+		       , secret, NSD_COOKIE_SECRET_SIZE);
+		nsd->cookie_count = 1;
```
```suggestion
		nsd->cookie_count = 1;
		memset(secret, 0, NSD_COOKIE_SECRET_SIZE);
```
The erase also happens later, so here too.

wcawijngaards commented on util.c:
```diff
@@ -1234,3 +1235,58 @@ int set_cpu_affinity(cpuset_t *set)
+	       , sizeof(struct cookie_secret) * (NSD_COOKIE_HISTORY_SIZE - 1));
+	memcpy( nsd->cookie_secrets[nsd->cookie_count - 1].cookie_secret
+	      , active_secret, NSD_COOKIE_SECRET_SIZE);
+	memset(active_secret, 0, NSD_COOKIE_SECRET_SIZE);
```
These kinds of zero memsets get removed by the compiler. Use `explicit_bzero` to make sure it does not get removed. Also for the other memset operations on the cookies.

wcawijngaards commented on remote.c:
```diff
@@ -2163,6 +2163,150 @@ do_del_tsig(RES* ssl, xfrd_state_type* xfrd, char* arg) {
+		secret_hex[NSD_COOKIE_SECRET_SIZE * 2] = '\0';
+		fprintf(f, "%s\n", secret_hex);
+	}
+	fclose(f);
```
If you want to memset the array, the 'secret_hex' array needs to be cleared here.

wtoorop replied:
resolved in aa3bca3

wcawijngaards commented on remote.c:
```diff
@@ -2163,6 +2163,150 @@ do_del_tsig(RES* ssl, xfrd_state_type* xfrd, char* arg) {
+	if(hex_pton(arg, secret, NSD_COOKIE_SECRET_SIZE) != NSD_COOKIE_SECRET_SIZE ) {
+		(void)ssl_printf(ssl, "invalid cookie secret: parse error\n");
+		(void)ssl_printf(ssl, "please provide a 128bit hex encoded secret\n");
+		memset(arg, 0xdd, strlen(arg));
```
If you memset 'arg' you should also clear the 'secret' array, which contains partial results. Also for memset calls for arg further in this function, there should be a clear of the 'secret' array, as well, if you want to erase parts of that data.

wtoorop replied:
resolved in aa3bca3

wcawijngaards commented on remote.c:
```diff
@@ -2163,6 +2163,150 @@ do_del_tsig(RES* ssl, xfrd_state_type* xfrd, char* arg) {
+			(void)ssl_printf(ssl, "staging: %s\n",  secret_hex);
+		else
+			(void)ssl_printf(ssl, "staging[%d]: %s\n", i, secret_hex);
+	}
```
If you want to clear the array, here `secret_hex` is the array to clear.

wtoorop replied:
resolved in aa3bca3

wcawijngaards commented on remote.c:
```diff
@@ -2226,6 +2370,14 @@ execute_cmd(struct daemon_remote* rc, RES* ssl, char* cmd, struct rc_state* rs)
 	} else if(cmdcmp(p, "del_tsig", 8)) {
 		do_del_tsig(ssl, rc->xfrd, skipwhite(p+8));
+	} else if(cmdcmp(p, "add_cookie_secret", 17)) {
+		do_add_cookie_secret(ssl, rc->xfrd, skipwhite(p+18));
```
```suggestion
		do_add_cookie_secret(ssl, rc->xfrd, skipwhite(p+17));
```
The string length is 17.

wcawijngaards commented on remote.c:
```diff
@@ -2226,6 +2370,14 @@ execute_cmd(struct daemon_remote* rc, RES* ssl, char* cmd, struct rc_state* rs)
+	} else if(cmdcmp(p, "print_cookie_secrets", 20)) {
+		do_print_cookie_secrets(ssl, rc->xfrd, skipwhite(p+20));
+	} else if(cmdcmp(p, "activate_cookie_secret", 22)) {
+		do_activate_cookie_secret(ssl, rc->xfrd, skipwhite(p+18));
```
```suggestion
		do_activate_cookie_secret(ssl, rc->xfrd, skipwhite(p+22));
```
The string length is 22.

wtoorop committed:
Apply suggestions from code review

Co-authored-by: Wouter Wijngaards <wcawijngaards@users.noreply.github.com>

wcawijngaards commented on remote.c:
```diff
@@ -2163,6 +2163,150 @@ do_del_tsig(RES* ssl, xfrd_state_type* xfrd, char* arg) {
+	for(size_t i = 0; i < nsd->cookie_count; i++) {
+		struct cookie_secret const* cs = &nsd->cookie_secrets[i];
+		ssize_t const len = hex_ntop(cs->cookie_secret, NSD_COOKIE_SECRET_SIZE,
+																 secret_hex, sizeof(secret_hex));
```
```suggestion
			secret_hex, sizeof(secret_hex));
```
Indent, nothing important.

wcawijngaards reviewed (approved):
Code looks good, thanks for already incorporating a bunch of changes!

wtoorop committed:
Rub out leftover secret buffers with explicit_bzero

wtoorop committed:
Update remote.c

Co-authored-by: Wouter Wijngaards <wcawijngaards@users.noreply.github.com>

wtoorop commented:
Thanks @wcawijngaards ! Maybe you'd like to rereview the last commit?

wtoorop committed:
Putting back previous version of install-sh

Sorry!  I should not have touched it, but I did somewhere in July 2019.  I've
tried to undo the touching with several git rebase attempts, but without
success. So now just putting back the previous version.

wtoorop committed:
Merge branch 'features/cookies' of github.com:NLnetLabs/nsd into features/cookies
